host/switchless_manager.c

```diff
@@ -0,0 +1,172 @@
+/* Copyright (c) Microsoft Corporation.  All rights reserved.
+ * Licensed under the MIT License. */
+
+#include "switchless_manager.h"
+#include <openenclave/internal/calls.h>
+#include "sgx/enclave.h"
+
+/* thread helper functions */
+#if _MSC_VER
+
+#define _thread_return_t DWORD WINAPI
+typedef LPVOID _thread_arg_type;
+#define _THREAD_RETURN_VAL 0
+
+typedef DWORD (*_thread_op_t)(LPVOID);
+
+static int _thread_create(
+    oe_thread_t* thread,
+    _thread_op_t op,
+    _thread_arg_t arg)
+{
+    *thread = CreateThread(NULL, 0, op, arg, 0, NULL);
+    return NULL == *thread;
+} /* _thread_create */
+
+static void _thread_join(_thread_arg_t thread)
+{
+    WaitForSingleObject(thread, INFINITE);
+    CloseHandle(thread);
+} /* _thread_join */
+
+#elif defined __GNUC__
+
+typedef void* _thread_return_t;
+typedef void* _thread_arg_t;
+#define _THREAD_RETURN_VAL NULL
+
+typedef void* (*_thread_op_t)(void*);
+
+static int _thread_create(
+    oe_thread_t* thread,
+    _thread_op_t op,
+    _thread_arg_t arg)
+{
+    return pthread_create(thread, NULL, op, arg);
+} /* _thread_create */
+
+static void _thread_join(oe_thread_t thread)
+{
+    pthread_join(thread, NULL);
+} /* _thread_join */
+
+#endif /* _MSC_VER or __GNUC__ */
+
+static _thread_return_t _enclave_worker_thread(_thread_arg_t arg)
+{
+    oe_enclave_t* enclave = (oe_enclave_t*)arg;
+
+    oe_ecall(
+        enclave,
+        OE_ECALL_LAUNCH_ENCLAVE_WORKER,
+        (uint64_t)(&enclave->switchless_manager.switchless),
+        NULL);
+
+    return _THREAD_RETURN_VAL;
+} /* _enclave_worker_thread */
+
+void oe_switchless_manager_init(oe_enclave_t* enclave)
+{
+    enclave->switchless_manager.switchless.state = OE_SWITCHLESS_STATE_STOPPED;
+    oe_lockless_queue_init(
+        &(enclave->switchless_manager.switchless.ecall_queue));
+} /* oe_switchless_manager_init */
+
+oe_result_t oe_switchless_manager_startup(oe_enclave_t* enclave)
+{
+    oe_result_t result = OE_OK;
+    oe_switchless_state_t state = OE_SWITCHLESS_STATE_STOPPED;
+
+    /* check the current state:
+       - if it is stopped then change to starting */
+#ifdef _MSC_VER
```

Comment:
I'd define the platform dependent atomic functions in a separate file, say `common/atomic.c`, and hide the `#ifdef _MSC_VER` and `#elif defined __GNUC__` there. I think it's much cleaner that way.

Reply:
The post condition for this method is not clear. It does not guarantee that when this function returns, the state is 'running'.
Say, in line 37, the state is running and thus the control does not go into line 37. Meanwhile another thread calls shutdown and changes the state to stopped.

Reply:
I agree that would be useful.  However, it would be a substantial amount of work to accomplish this.  I'm only using a few functions atomic functions (exchange, compare-exchange, load) and only for uint32_t types.  The Windows and GNU implementations for atomic intrinsic functions are implemented very differently from each other.  I would be nice if Windows supported stdatomic.h.  I do not believe that stdatomic.h is supported in enclaves at this point.
I have  encapsulated the usage of atomic functions into file scope functions to make the functional code more less distracting.